### ISSUE TITLE
Unit shadows aren't being drawn properly.

### ISSUE THREAD

**Issue by aeonios** [ZeroK-RTS/Zero-K#3734]
I have both unit and terrain shadows enabled.

I see shadows for terrain, smoke, and nanoframes show up properly, but as soon as a nanoframe is completed its shadow disappears. There seem to be some weird effects as the nanoframe approaches completion, although I can't tell if that's just the nanoframe doing its final fill-in or if the shadow actually starts to disappear as the nanoframe is replaced by the unit model.

I don't think explosion debris is casting shadows either, so it's probably an issue with the solid models not being drawn to the shadow map.

[infolog.txt](https://github.com/ZeroK-RTS/Zero-K/files/3700288/infolog.txt)

**Comment by Anarchid** [comment]
ZK has been drawing everything using with CUS shaders for a while (this is why you may notice some emissives pulsing etc). Possibly this fails for you?

Can you provide a screenshot of the issue for completeness?

**Comment by aeonios** [comment]
Sure.
![screen00017](https://user-images.githubusercontent.com/7027766/66400209-0a211680-e9af-11e9-9d96-198b53c41b54.jpg)

I've seen the glowing unit thing and that seems to work fine. I've also noticed some new projectile lights on flaming debris. :P Aside from weird screen flickering from duck and to a lesser extent fencer missiles I haven't really noticed anything else being off though.

**Comment by Anarchid** [comment]
Does the problem persist if you `/luarules disable` or `/luaui disable`?

**Comment by aeonios** [comment]
[infolog_full.txt](https://github.com/ZeroK-RTS/Zero-K/files/3707152/infolog_full.txt)
Hopefully a better infolog.

/luarules disable does fix the shadows. /luaui disable does not.

**Comment by sprunk** [comment]
@aeonios you can try the `bug3734wa` springsetting workaround, see #3739.

**Comment by aeonios** [comment]
No, this workaround does not work. I've had it in my local zk $VERSION the whole time without realizing it and I still don't have shadows.

**Comment by sprunk** [comment]
@lhog ^

**Comment by lhog** [comment]
Also with disabled custom shadow map depth shader you should see terrain shadows casted by units. Do you see those?

**Comment by aeonios** [comment]
Yes, commenting out https://github.com/lhog/Zero-K/blob/master/LuaRules/Gadgets/api_custom_unit_shaders.lua#L565 does fix shadows. If https://github.com/ZeroK-RTS/Zero-K/blob/master/ModelMaterials/Templates/defaultMaterialTemplate.lua#L729 is affecting shadow map drawing in any way then that's all on you. It shouldn't, period. It probably doesn't, either.

'Recently' is like since 2 years ago. BAR shadows worked fine back then, and BYAR shadows still work fine now. I don't understand why you're using a shadow transform that is different from spring's default shader though. Obviously that is the problem.

**Comment by lhog** [comment]
If manual commenting out did the trick, then it means you probably didn't apply the workaround 
setting the way you should.

Did you create 'bug3734wa=1' entry in the springsettings.cfg ? Did the setting entry appear in the infolog?

'shadowsQuality' affects the shadow mapping perceptual quality, but with your recent finding in mind it is apparently not related to the issue you have.

**Comment by lhog** [comment]
> I don't understand why you're using a shadow transform that is different from spring's default shader though. Obviously that is the problem.

Elaborate what you mean here?
Also obviously it is not the problem, because it works for everyone else but you.

**Comment by aeonios** [comment]
> Did you create 'bug3734wa=1' entry in the springsettings.cfg ?
No. Why should I have to? You can't expect any regular user to do that.

>Also obviously it is not the problem, because it works for everyone else but you.
No, I'm the only one who has the problem AND noticed it, AND is competent enough to actually help debug it. Everyone else who has the problem you just never hear about it, except that one guy because he happened to post a video, and he didn't even notice that there were no shadows, googlefrog just happened to notice it in the video.

Also shadows work fine when you copy the code from the default shader in spring, which is to say that that code is working and yours isn't. Like, can it possibly get any more specific than that? No other explanation even makes sense.

Speaking of which, how does spring handle shadows for smoke particles? Those are ostensibly translucent (as opposed to the 1-bit alpha trees) and that was working fine. Using a horrible hack workaround for something that can be fixed without one is not a solution at all. Especially if the workaround requires manual user intervention.

**Comment by lhog** [comment]
> No. Why should I have to? You can't expect any regular user to do that.

I have no means to determine if your systems or anyone's else belong to modern or potato bin. Therefore it's an individual setting you have to do yourself. Given your tech savvy background, I expected you would read the PR code, before making reports that it didn't work.
However you expressed a good point: For the regular users we should add a chobby checkbox for the W/A setting.

> No, I'm the only one who has the problem AND noticed it, AND is competent enough to actually help debug it. Everyone else who has the problem you just never hear about it, except that one guy because he happened to post a video, and he didn't even notice that there were no shadows, googlefrog just happened to notice it in the video.

I can agree that you are not alone, however you can't deny that there are only a few folks with shadow issues. I'm aware of exactly three cases out of 1000+ active players.


> Also shadows work fine when you copy the code from the default shader in spring, which is to say that that code is working and yours isn't. Like, can it possibly get any more specific than that? No other explanation even makes sense.

We have been through this before in the Discord chat. Surprise me it is not a dejavu of the following: you pasted some code -> it failed compilation -> you failed to notice that it failed to compile -> the engine fallback kicked in.

Do you realize that your claims that my code does not work or uses some wrong "transforms" do not pass the reality check? If it works for everyone, but a selected few, it cannot be wrong or incorrect. The only "wrong" part there is the 10 years old GLSL version tag, which happens to cause bugs on your end. If you could just update your GPU you would witness how the issue fixes itself on the bug-free drivers. The most I could do for your ancient kit was to issue a manual override workaround, which I did. Please either update your gear or apply the proposed W/A.



> Speaking of which, how does spring handle shadows for smoke particles? Those are ostensibly translucent (as opposed to the 1-bit alpha trees) and that was working fine.

It's handled no different than other game object types: the in-shader alpha check discards fragments with less than 0.5 value.



> Using a horrible hack workaround for something that can be fixed without one is not a solution at all. Especially if the workaround requires manual user intervention.

The horrible workaround exists because of your horrible drivers. We tried the other ways to make the shadow pass shader work on `#version 150 compatibility`, unfortunately it did nothing. I asked if you could visualize `$shadow` texture in attempt to see what is going on there, but you didn't do that. What else can I do?

**Comment by sprunk** [comment]
Drivers issue, workaround exists. See https://github.com/ZeroK-RTS/Chobby/issues/719 about exposing the option for non-techsavvy people.

**Comment by CHmuc** [comment]
Umm,..
Sorry to write about this topic again.
But I have the same problem withe the missing shadows.
I see them while unit is building. As soon as it´s finished, the shadow is gone.
I _can_ get them back by turning off "shiny units", but that looks bad as well.

I would like "shiny units" _and_ shadows.
I use a Radeon RX 580 with drivers dated this month.

Is this possible?

My logfile
[infolog.txt](https://github.com/ZeroK-RTS/Zero-K/files/4500276/infolog.txt)

**Comment by lhog** [comment]
See https://github.com/ZeroK-RTS/Zero-K/pull/3739#issuecomment-616225669

TLDR: the cause is known, the fix too. Just need to port fix.
